오픈마켓1 [STEP 2] 휴, 키위

Hugh-github opened the pull request:
@jryoun1
안녕하세요 제이크!!
FlowLayout을 이용해  STEP2를 다시 작성해 봤습니다.
STEP2 리뷰 추가로 해주셔서 정말 감사합니다 😭

### 구현한 기능
+ UICollectionViewDelegateFlowLayout을 채택해 collectionview의 layout 상태에 따라 다른 itemSize를 반환하도록 구현했습니다.
+ UIImageView를 확장해 비동기 적으로 Data를 가져와 view를 설정하도록 했습니다.

### 궁금한 점
+ 현재 저희가 List와 Grid에서 사용할 cell을 각각 만들어 주었기 때문에 세그먼트를 통해 layout을 변경하는 과정에서 reloadItem를 사용하지 않으면 화면이 전화되는 시점에 존재하는 cell들의 layout이 변하지 않는 걸 확인했습니다. reloadItem를 통해 cell을 새로 만들어 주니 해당 문제는 해결됐지만 모든 indexPath에 대해 cell을 만들어 주는 것이 비효율적이라 생각합니다. 혹시 어떠한 방법으로 해당 문제를 해결할 수 있을지 궁금합니다.

+ 스크롤을 통한 Pagination 과정에서 fetchData를 호출할 때 reloadData를 호출하여 collectionView의 전체적인 업데이트를 해주었습니다. 추가적으로 cell에 대한 update를 해주기 위해 reconfigureItems를 사용해 기존 cell을 유지하면서 업데이트를 하려 했지만 에러(특정 경우 앱이 꺼짐 및 layout 전환 시 몇몇 cell의 layout이 변하지 않음) 가 발생해 결국 모든 indexPath에 대해 reloadItem을 해주었습니다. 혹시 reconfigureItems를 사용하면 발생하는 에러가 reconfigureItems는 추가된 cell이 아니라 기존 cell에 대한 데이터만 업데이트를 하기 때문에 발생한 에러인지 궁금합니다.

+ 네트워크 기반의 데이터를 가지고 올때 Data(contentsOf:)이니셜라이저를 사용하지 말라는 아래의 사진과 같은 글을 공식문서에서 보았습니다. 
![](https://i.imgur.com/fw5Jvye.png)

하지만 해당 이니셜라이저를 메인 스레드가 아닌곳에서는 사용을 해도 된다는 글을 보아 혹시 이에 대해 제이크는 어떻게 생각하시는지 궁금하여 질문 드립니다!

2.stackoverflow
![](https://i.imgur.com/JQ7D112.png)


+ UIImageView를 확장해 view를 설정했습니다.
```swift
DispatchQueue.global().async {
    guard let url = URL(string: item.thumbnail),
          let imageData = try? Data(contentsOf: url) else { return }
    DispatchQueue.main.async {
        self.thumbnailView.image = UIImage(data: imageData)
    }
}
```
위 코드와 같은 방식으로도 동일한 효과를 얻을 수 있다고 생각하는데 현업에서 서버에서 받아온 image를 view에 사용하기 위해 어떠한 방식을 사용하는지 궁금합니다.
추가로 현재 cell을 생성할 때 여러 이미지를 처리하는 게 아니기 때문에 `Data(contentsOf:)`를 비동기로 처리 안 해도 된다고 생각하는데 하나의 이미지를 처리할 때도 비동기로 처리하는 게 맞는지 궁금합니다.

+ 캐싱과 관련해서 어떤 data들을 캐시로 저장해 사용하는 게 좋을지 궁금합니다. 실제로 당근마켓과 같은 앱들은 상품마다 이미지가 다르기 때문에 굳이 이미지를 캐싱해서 사용해야 하는지 잘 모르겠습니다. 혹시 제이크라면 오픈마켓에서 어떤 data들을 캐싱해서 사용할 지 궁금합니다.

Hugh-github committed:
FlowLayout 1차 완성본(추가 수정 예정)

kiwi1023 committed:
feat: UIImageView 확장 메서드 구현

Hugh-github committed:
feat: NumberFormatter 추가

kiwi1023 committed:
refactor: Main 스토리 보드 삭제

Hugh-github committed:
refactor: 프로젝트 작성자 이름 변경

kiwi1023 committed:
docs: 리드미 수정

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+    private var currentPage = 1
+    private let itemsPerPage = 20
+    private var items = [Item]()
+    private var isPageRfreshing: Bool = true
```
오타가 있네요 🙂
위에서 다른 프로퍼티들과는 다르게 타입 명시를 해준 이유가 있을까요? 🙋‍♂️

Hugh-github replied:
타입 명시하는 걸 일치 시켰어야 했는데 확인하지 못했습니다. 해당 부분은 isPageRefreshing의 타입 명시를 제거해 주었습니다.

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+                
+                DispatchQueue.main.async {
+                    if self.currentPage == 1 {
+                        self.configureListCollectionView()
```
collectionView의 설정은 네트워크의 성공과는 무관하게 항상 우선되어야할 것 같은데 여기서 진행되는 이유가 있을까요? 🙋‍♂️

Hugh-github replied:
configureCollectionView를 ViewDidLoad에서 호출하고 fetchData 내부에서는 reloadData를 통해 collectionView를 update하도록 수정했습니다

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        return listLayout
+    }
+    
+    private func configureListCollectionView() {
```
함수명은 ListCollectionView에 대해서 설정해줄 것 같은 함수명인데, 
실제 함수 내부에서는 collectionView를 설정해주고 있는 것 같네요 🤔

kiwi1023 replied:
함수명 변경해주었습니다~

jryoun1 commented on OpenMarket/OpenMarket/Model/Extension/Int+extension.swift:
```diff
@@ -0,0 +1,18 @@
+        let numberFormatter = NumberFormatter()
+        numberFormatter.numberStyle = .decimal
+        
+        guard let number = numberFormatter.string(for: self) else { return "" }
```
변환에 실패했을 때 빈 문자열을 반환 해준다면, 서버에서 빈 값이 내려왔을 때와 제대로 내려왔지만 반환에 실패한 경우에 결과가 동일하기 때문에 어디가 문제인지 찾기 어려울 것 같네요.
따라서 Optional 타입을 반환해 실패한 경우에는 nil을 반환하는 것은 어떨까요?

kiwi1023 replied:
수정하였습니다!

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+            case .success(let responseData):
+                guard let data = responseData,
+                      let itemData: ItemList = JSONDecoder.decodeJson(jsonData: data) else { return }
+                itemData.pages.forEach { item in
```
append(contentsOf) 를 사용해서 하나씩 추가하지 않고 한 번에 추가할 수도 있을 것 같아요. 🙏🏻

Hugh-github replied:
해당 메서드를 이용해 수정해 주었습니다  ☺️

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        switch sender.selectedSegmentIndex {
+        case 0:
+            collectionView.setCollectionViewLayout(createListLayout(), animated: true)
+            DispatchQueue.main.async { [self] in
```
여기가 말씀하신 부분이군요. 
> 현재 저희가 List와 Grid에서 사용할 cell을 각각 만들어 주었기 때문에 세그먼트를 통해 layout을 변경하는 과정에서 reloadItem를 사용하지 않으면 화면이 전화되는 시점에 존재하는 cell들의 layout이 변하지 않는 걸 확인했습니다. reloadItem를 통해 cell을 새로 만들어 주니 해당 문제는 해결됐지만 모든 indexPath에 대해 cell을 만들어 주는 것이 비효율적이라 생각합니다. 혹시 어떠한 방법으로 해당 문제를 해결할 수 있을지 궁금합니다.

2가지 방법이 있을 것 같습니다.
1. flowLayout을 변경해주기 전에 collectionView.reloadData()를 호출해서 아예 처음부터 그려줄 수 있도록 하는 방법 
```swift
collectionView.reloadData()
collectionView.setCollectionViewLayout(createListLayout(), animated: true)
```
그러나 이 방법은 list, grid에서 보고 있던 page이 초기화되기 때문에 기획에 따라서 결정할 수 있을 것 같네요. 
만약 기획이 list -> grid로 넘어갈때 보고 있던 상품을 보여줘야한다라면 이 방법이 아닌 다른 방법을 사용해야할 것 같습니다.


2. 현재 화면에 보이고 있는 item들에 대해서만 reload를 진행해주기
```swift
collectionView.setCollectionViewLayout(createListLayout(), animated: true)
let visibleItems = collectionView.indexPathsForVisibleItems
collectionView.reloadItems(at: visibleItems)
```
현재 두 분이 작성하신 코드는 모든 item들을 다 reload해주고 있었다면 collectionView의 [indexPathsForVisibleItems](https://developer.apple.com/documentation/uikit/uicollectionview/1618020-indexpathsforvisibleitems)를 활용해 실제 화면에 보여지고 있는 cell들의 item만 reload하도록 변경할 수 있을 것 같네요.

Hugh-github replied:
reloadData를 사용하게 되면 제이크의 말처럼 page가 초기화되기 때문에 layout전환시 화면이 맨 위로 올라가져 있는걸 확인했습니다. 따라서 저희는 reloadItem을 사용해 보여지고 있는 cell들의 item만 reload 해주었습니다.

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+                    }
+                    self.collectionView.reloadData()
+                    let indexPath = Array(0..<(self.currentPage * self.itemsPerPage)).map { IndexPath(item: $0, section: 0) }
+                    self.collectionView.reloadItems(at: indexPath)
```
위에서 이미 reloadData를 호출하는데 reloadItems가 또 호출되는 이유가 있을까요? 🤔

Hugh-github replied:
해당 문제 같은 경우 collectionView 생성을 ViewDidLoad에서 해주고 reloadData를 사용하니 reloadItem을 사용하지 않아도 문제가 없음을 확인했습니다.
저희가 처음에 reloadItem을 사용한 이유 같은 경우 layout 전환 시 현재 보이는 cell의 layout을 다른 화면에서 그대로 사용하기 때문에 해당 cell을 다시 만드는 방법을 사용했습니다.(이 경우에는 reloadData를 제거해 주어도 정상적으로 동작합니다.)

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+                    let indexPath = Array(0..<(self.currentPage * self.itemsPerPage)).map { IndexPath(item: $0, section: 0) }
+                    self.collectionView.reloadItems(at: indexPath)
+                    self.activityIndicatorView.stopAnimating()
+                    self.isPageRfreshing = true
```
이 작업은 main스레드에서 실행되어야하는 작업인가요?

Hugh-github replied:
isPageRefreshing을 true로 바꿔 주는 작업은 main 스레드에서 하지 않아도 된다고 생각합니다. 하지만 collectionView item을 reload하는 작업 같은 경우 UI Update와 관련 있다고 생각하여 main 스레드에서 해주었습니다.

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+                    self.activityIndicatorView.stopAnimating()
+                    self.isPageRfreshing = true
+                }
+            default:
```
실패한 케이스에 대한 처리는 없을까요? 🤔

kiwi1023 replied:
에러처리추가 하였습니다!

jryoun1 commented on OpenMarket/OpenMarket/View/GridFlowLayout.swift:
```diff
@@ -0,0 +1,24 @@
+
+import UIKit
+
+class GridFlowLayout: UICollectionViewFlowLayout {
```
따로 클래스로 만든 이유가 있을까요? 🙋‍♂️
(추가로 상속될 일이 없다면 final 키워드를 붙여주세요)

kiwi1023 replied:
키워드 추가 해주었습니다!!

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        guard collectionView.contentSize.height > 0 else { return }
+        if collectionView.contentOffset.y > (collectionView.contentSize.height - collectionView.frame.height) {
+            if isPageRfreshing {
+                self.currentPage += 1
```
사용자가 계속 스크롤링할 때 마지막 페이지인 경우에도 +1을 하고 서버로 요청을 보낼 것 같은데 이에 대한 처리가 필요해보입니다. 
(api response를 보면 has_next 라는 필드가 있네요)

Hugh-github replied:
itemData의 hasNext가 true인 경우에만 isPageRefreshing이 true가 되도록 조건문을 추가해 주었습니다.

jryoun1 replied:
추후에 스텝을 다 마친다면 보너스로 페이징을 요청할 때에도 loading indicator를 보여줄 수 있으면 더욱 좋은 UX가 될 것 같습니다. 

(다음 페이지를 로드하는 동안 로드중임을 알 수 있고, 만약 마지막 페이지인 경우에는 loading indicator가 보이지 않으므로 사용자는 모든 페이지를 다 봤다고 알 수도 있을 것 같습니다)

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+//MARK: DataSource
+extension MainViewController: UICollectionViewDataSource {
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return currentPage * itemsPerPage
```
items의 개수가 아닌 두 값을 곱한 결과를 사용하고 있는 이유가 있을까요? 🙋‍♂️

Hugh-github replied:
items.count를 이용해 item 수를 반환해 주었습니다.

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+            let widthOfCells = collectionView.bounds.width - (collectionView.contentInset.left + collectionView.contentInset.right)
+            let widthOfSpacing = CGFloat(gridLayout.numberOfColumns - 1) * gridLayout.minimumInteritemSpacing
+            let width = (widthOfCells - widthOfSpacing) / CGFloat(gridLayout.numberOfColumns)
+            let height = collectionView.bounds.height * 0.35
```
itemSize에 대한 설정도 미리 해줄 수 있을 것 같아보입니다. 🙏🏻
클래스로 만들지 않고, vc에서 아래와 같이 선언이 가능할 것 같은데 GridFlowLayout 클래스를 따로 만드신 이유가 궁금합니다. 

```swift
    private var gridLayout: UICollectionViewFlowLayout = {
        let layout = UICollectionViewFlowLayout()
        layout.scrollDirection = .vertical
        layout.minimumLineSpacing = 5
        layout.minimumInteritemSpacing = 5
        layout.itemSize = ...
        return layout
    }()
```

Hugh-github replied:
현재 GridLayout에서 columns 속성을 가지고 ItemSize를 결정하기 위해 Custom 해서 만들었습니다. 또한 UICollectionViewDelegateFlowLayout을 통해 itemSize를 결정하는 과정에서 collectionView의 layout이 GridFlowLayout 인지를 확인해 itemSize를 결정하는 방법이 더 좋다고 생각했습니다.

jryoun1 reviewed (changes requested):
@Hugh-github @kiwi1023 
이미 진행하신 step에 있어서도 다른 방법을 공부해서 다시 코드를 작성해보신 건 정말 대단합니다 👏
collectionView는 정말 중요하기 때문에 좋은 경험이 될 것이라고 생각합니다. 👍🏻

몇 가지 오류들이 있어서 먼저 말씀드립니다.
1. paging이 제대로 되지 않고 있습니다. 
> dataCorrupted(Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "pages", intValue: nil), _JSONKey(stringValue: "Index 3", intValue: 3), CodingKeys(stringValue: "price", intValue: nil)], debugDescription: "Parsed JSON number <0.78> does not fit in Int.", underlyingError: nil))

2. List에서 페이징을 시도하다가 페이징이 동작하지 않는 상황에서 Grid를 눌렀을 때 아래 에러와 함께 크래시가 발생합니다.  (Grid -> List도 동일하게 문제 발생)
> Thread 1: Fatal error: Index out of range
```swift
// line 140 부터 이고, 크래시는 143번째 라인 
            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "ListCell", for: indexPath)
                    as? ListCollectionViewCell else { return UICollectionViewCell() }
            cell.resetContent()
            cell.configureContent(item: items[indexPath.row])
            return cell

// line 147 부터 이고, 크래시는 150번째 라인 
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "GridCell", for: indexPath)
                as? GridCollectionViewCell else { return UICollectionViewCell() }
        cell.resetContent()
        cell.configureContent(item: items[indexPath.row])
        return cell
```

3. auto-layout 이 제대로 잡히지 않은 것 같네요.
<img width="476" alt="image" src="https://user-images.githubusercontent.com/45090197/181799762-c5362d13-d39a-4b04-a3e9-785f07ea2801.png">

## 궁금한 점에 대한 답변
> 현재 저희가 List와 Grid에서 사용할 cell을 각각 만들어 주었기 때문에 세그먼트를 통해 layout을 변경하는 과정에서 reloadItem를 사용하지 않으면 화면이 전화되는 시점에 존재하는 cell들의 layout이 변하지 않는 걸 확인했습니다. reloadItem를 통해 cell을 새로 만들어 주니 해당 문제는 해결됐지만 모든 indexPath에 대해 cell을 만들어 주는 것이 비효율적이라 생각합니다. 혹시 어떠한 방법으로 해당 문제를 해결할 수 있을지 궁금합니다.

아래 코멘트에 답변 남겨두었습니다. 

> 스크롤을 통한 Pagination 과정에서 fetchData를 호출할 때 reloadData를 호출하여 collectionView의 전체적인 업데이트를 해주었습니다. 추가적으로 cell에 대한 update를 해주기 위해 reconfigureItems를 사용해 기존 cell을 유지하면서 업데이트를 하려 했지만 에러(특정 경우 앱이 꺼짐 및 layout 전환 시 몇몇 cell의 layout이 변하지 않음) 가 발생해 결국 모든 indexPath에 대해 reloadItem을 해주었습니다. 혹시 reconfigureItems를 사용하면 발생하는 에러가 reconfigureItems는 추가된 cell이 아니라 기존 cell에 대한 데이터만 업데이트를 하기 때문에 발생한 에러인지 궁금합니다.

어떠한 코드에서 에러가 발생했는지 글로 남겨주신 것만으로는 확인하기 어려울 것 같습니다. 
추가로 reloadData를 해줬는데 reconfigureItems를 호출하는 이유가 무엇일까요? 🤔 

> 하지만 해당 이니셜라이저를 메인 스레드가 아닌곳에서는 사용을 해도 된다는 글을 보아 혹시 이에 대해 제이크는 어떻게 생각하시는지 궁금하여 질문 드립니다! 

아래에 추가로 남겨주신 것도 포함해서 답변드립니다. 🙏🏻
해당 메서드는 **동기적으로 동작**합니다. 동기적으로 동작한다는 것은 모든 스레드 작업이 멈추고 해당 작업이 끝날 때까지 다른 작업들이 전부 멈춰지게 됩니다. 이는 사용성에 있어서 매우 좋지 않기 때문에 해당 메소드를 사용해서 remote에 있는 데이터를 받아오는 것은 바람직하지 않습니다. 

추가로 DispatchQueue.global().async로 감싸준다고 했을 때 일단 메인 스레드를 멈추는 문제는 해결이 되지만, 여전히 GCD에서 해당 작업을 할당받는 스레드가 동기적으로 동작하게 되는 문제가 발생되고, 이는 퍼포먼스 측면에서 이슈가 있을 수 있습니다. 따라서 이 방법도 권장하지 않습니다.

또한 Data(contentsOf:)를 사용하는 경우에는 네트워크의 오류인지, HTTP 오류인지 판단할 수 있는 URLSession과는 다르게 판단이 불가능하기 때문에 역시나 권장하지 않습니다.

현업에서는 kingfisher라는 외부 라이브러리를 사용하거나, URLSession을 사용해서 직접 구현해서 사용합니다. 지금 단계에서는 URLSession을 사용해서 구현해보시는 걸 추천드립니다. 

> 추가로 현재 cell을 생성할 때 여러 이미지를 처리하는 게 아니기 때문에 Data(contentsOf:)를 비동기로 처리 안 해도 된다고 생각하는데 하나의 이미지를 처리할 때도 비동기로 처리하는 게 맞는지 궁금합니다.

동기와 비동기의 개념에 대해서 좀 더 공부를 해보시는걸 추천드립니다. 🙏🏻 
만약 이미지의 크기가 크고 다운받는데 시간이 오래 걸리는데 이미지를 다운받는 로직을 동기로 수행하게 된다면 그 동안 다른 작업들은 전혀 진행하지 못하게 됩니다. 반면 해당 작업을 비동기로 처리하게 된다면, 이미지를 다운 받는 동안 앱내에서는 다른 처리들을 진행하다가 이미지가 다운이 되는 시점에 다시 completionHandler를 통해서 처리를 해주면 되므로 성능면에서 차이가 많이 발생하게 될 것 같네요. 

> 캐싱과 관련해서 어떤 data들을 캐시로 저장해 사용하는 게 좋을지 궁금합니다. 실제로 당근마켓과 같은 앱들은 상품마다 이미지가 다르기 때문에 굳이 이미지를 캐싱해서 사용해야 하는지 잘 모르겠습니다. 혹시 제이크라면 오픈마켓에서 어떤 data들을 캐싱해서 사용할 지 궁금합니다.

캐싱을 하는 이유가 무엇일까요? 
만약 캐싱을 하지 않는 다면 이미지의 경우에는 매번 다운을 받아야할 것입니다. 
예를 들어서 지금은 list, grid만 보여주는 화면이 있어 캐싱이 의미가 없지만, 상품 상세 혹은 상품 작성과 같은 화면을 갔다가 돌아왔을 때 캐싱이 되어있지 않다면 다시 모든 이미지를 다운받아야 합니다. 
반면 캐싱이 되어있다면 앱이 맨 처음 실행되어 해당 이미지를 처음으로 다운로드 하지 않는 이상 그 이후에는 캐싱된 이미지를 사용하기 때문에 성능면에서 차이가 많이 날 것 같네요.

Hugh-github committed:
refactor: 오타 수정 및 불필요한 함수 호출 제거

- isPageRefreshing 오타 수정
- reloadData 제거
- indexPathsForVisibleItems를 이용해 현재 보이는 cell만 reloadItem 진행
- 반환하는 item 수를 items.count로 설정

kiwi1023 committed:
refactor: 네이밍 수정 및 접근제한자 수정

Hugh-github committed:
refactor: 네트워크와 관련없이 collectionView가 생성되도록 수정, itemData의 hasNext가 ture일 경우에만 pagination되도록 수정

kiwi1023 committed:
refactor: 에러처리 추가

Hugh-github commented:
pagination과 오토레이 아웃 같은 경우 어제 서버에 문제가 있어 Item타입의 price, discountedPrice, barginPrice의 타입을 Double 형태로 바꾸어 해당 코드를 실행하면 정상적으로 동작하는 걸 확인했습니다.

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        case 0:
+            DispatchQueue.main.async { [self] in
+                collectionView.setCollectionViewLayout(createListLayout(), animated: true)
+                let indexPath = collectionView.indexPathsForVisibleItems
```
indexPath보다 좀 더 명확한 변수명이면 좋을 것 같네요 🙏🏻

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+                collectionView.reloadItems(at: indexPath)
+            }
+        case 1:
+            DispatchQueue.main.async { [self] in
```
GCD에서는 weak self를 사용하지 않아도 될까요? 🙋‍♂️

jryoun1 commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,45 @@
+        let secondQuery = URLQueryItem(name: OpenMarketURL.itemsPerPageQuery.description, value: "\(itemsPerPage)")
+        baseComponents?.queryItems = [firstQuery, secondQuery]
+        
+        guard let requestURL = baseComponents?.url else { return }
```
여기서도 error처리가 필요해보입니다.

jryoun1 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,110 @@
+    }
+    
+    func configureContent(item: Item) {
+        guard let stock = item.stock.formatNumber() else { return }
```
stock이 없다면 여기서 early return되어 stock을 제외한 다른 내용들도 설정이 되지 않을 것 같은데 if let을 사용해야할 것 같습니다.

jryoun1 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,110 @@
+            priceLabel.textColor = .systemRed
+            priceLabel.isHidden = false
+            
+            guard let priceText = priceLabel.text else { return }
```
여기도 동일합니다.

jryoun1 commented on OpenMarket/OpenMarket/View/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,149 @@
+    }
+    
+    func configureContent(item: Item) {
+        guard let stock = item.stock.formatNumber() else { return }
```
동일합니다.

jryoun1 commented on OpenMarket/OpenMarket/View/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,149 @@
+            priceLabel.isHidden = false
+            priceLabel.textColor = .systemRed
+            
+            guard let priceText = priceLabel.text else { return }
```
여기도!!

jryoun1 commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,182 @@
+    @objc private func tapSegment(sender: UISegmentedControl) {
+        switch sender.selectedSegmentIndex {
+        case 0:
+            DispatchQueue.main.async { [self] in
+                collectionView.setCollectionViewLayout(createListLayout(), animated: true)
+                let indexPath = collectionView.indexPathsForVisibleItems
+                collectionView.reloadItems(at: indexPath)
+            }
+        case 1:
+            DispatchQueue.main.async { [self] in
+                collectionView.setCollectionViewLayout(gridLayout, animated: true)
+                let indexPath = collectionView.indexPathsForVisibleItems
+                collectionView.reloadItems(at: indexPath)
+            }
```
해당 부분으로 인해서 grid, list를 반복하는 경우에 cell들이 계속해서 생성되고 있는 것 같습니다. 

처음에 list일 때 viewHierarchy를 확인해보면 레이아웃의 문제가 없는데, 다시 grid로 변경후에 viewHierarchy를 확인해보면 기존에 ListCollectionViewCell이 남아있고 여기서 layout이 변경되면서 해당 cell의 layout이 모호하다는 에러가 발생하네요. 

추가로 list -> grid, grid -> list를 여러번 반복하면 계속해서 cell들이 계속해서 생성되면서 위에 중첩되고 있네요. 
<img width="1200" alt="image" src="https://user-images.githubusercontent.com/45090197/182154680-91198c11-f1d6-4890-9eb0-4fdb894b288c.png">

원인은 `reloadItems(at:)`의 경우에는 기존에 cell들을 버리고 cell을 다시 생성해서 보여주기 때문에 그렇다고 합니다.
해결책 중 하나는 reloadData를 통해서 모든 cell을 버리고 다시 그려주는 방법을 사용할 수 있습니다. 다만 이 경우에는 기존에 보고 있던 위치로 이동하기 위해서 추가적인 로직이 필요할 것 같습니다! 

해당 로직은 indexPathsForVisibleItems를 사용해서 한 번 구현해보시면 좋을 것 같네요! 
hint: collectionView의 scrollToItem 함수도 같이 사용해보시면 될 것 같아요. 

reloadData를 사용하지 않고도 cell들이 계속해서 중첩되어 생기는 문제를 해결할 수 있는 다른 추가적인 방법에 대해서는 좀 더 알아봐야할 것 같네요 🤔

jryoun1 reviewed (approved):
@Hugh-github @kiwi1023 
주말에도 수정하시느냐고 고생하셨습니다. 👍🏻

대부분 잘 수정해주셨고, layout 관련해서 확인하는 도중에 오류를 발견해서 
해결방법에 대해서 코멘트 남겨드립니다. 다른 해결 방법에 대해서도 좀 더 알아봐야 할 것 같네요 🤔

수정할 것이 많지 않아서 다음 스텝 진행하시면서 수정하셔도 될 것 같아서 이번 스텝은 여기서 머지하겠습니다!
